stateproof/verify/stateproof.go

```diff
@@ -0,0 +1,185 @@
+			total, proto.StateProofInterval, latestRoundInProof, firstValid, provenWeight, scaledWeight)
+		return 0
+	}
+
+	return w
+}
+
+// GetProvenWeight computes the parameters for building or verifying
+// a state proof for the interval (votersHdr, latestRoundInProofHdr], using voters from block votersHdr.
+func GetProvenWeight(votersHdr bookkeeping.BlockHeader, latestRoundInProofHdr bookkeeping.BlockHeader) (uint64, error) {
+	proto := config.Consensus[votersHdr.CurrentProtocol]
+
+	if proto.StateProofInterval == 0 {
+		return 0, errStateProofNotEnabled
+	}
+
+	if votersHdr.Round%basics.Round(proto.StateProofInterval) != 0 {
+		err := fmt.Errorf("votersHdr %d not a multiple of %d",
+			votersHdr.Round, proto.StateProofInterval)
+		return 0, err
+	}
+
+	if latestRoundInProofHdr.Round != votersHdr.Round+basics.Round(proto.StateProofInterval) {
+		err := fmt.Errorf("certifying block %d not %d ahead of voters %d",
+			latestRoundInProofHdr.Round, proto.StateProofInterval, votersHdr.Round)
+		return 0, err
+	}
+
+	totalWeight := votersHdr.StateProofTracking[protocol.StateProofBasic].StateProofOnlineTotalWeight.ToUint64()
+	provenWeight, overflowed := basics.Muldiv(totalWeight, uint64(proto.StateProofWeightThreshold), 1<<32)
+	if overflowed {
+		err := fmt.Errorf("overflow computing provenWeight[%d]: %d * %d / (1<<32)",
+			latestRoundInProofHdr.Round, totalWeight, proto.StateProofWeightThreshold)
+		return 0, err
+	}
+
+	return provenWeight, nil
+}
+
+// ValidateStateProof checks that a state proof is valid.
+func ValidateStateProof(latestRoundInIntervalHdr *bookkeeping.BlockHeader, stateProof *stateproof.StateProof, votersHdr *bookkeeping.BlockHeader, nextStateProofRnd basics.Round, atRound basics.Round, msg *stateproofmsg.Message) error {
+	proto := config.Consensus[latestRoundInIntervalHdr.CurrentProtocol]
+
+	if proto.StateProofInterval == 0 {
+		return fmt.Errorf("rounds = %d: %w", proto.StateProofInterval, errStateProofNotEnabled)
+	}
+
+	if latestRoundInIntervalHdr.Round%basics.Round(proto.StateProofInterval) != 0 {
+		return fmt.Errorf("state proof at %d for non-multiple of %d: %w", latestRoundInIntervalHdr.Round, proto.StateProofInterval, errNotAtRightMultiple)
+	}
+
+	votersRound := latestRoundInIntervalHdr.Round.SubSaturate(basics.Round(proto.StateProofInterval))
+	if votersRound != votersHdr.Round {
+		return fmt.Errorf("new state proof is for %d (voters %d), but votersHdr from %d: %w",
+			latestRoundInIntervalHdr.Round, votersRound, votersHdr.Round, errInvalidVotersRound)
+	}
+
+	if nextStateProofRnd == 0 || nextStateProofRnd != latestRoundInIntervalHdr.Round {
+		return fmt.Errorf("expecting state proof for %d, but new state proof is for %d (voters %d):%w",
+			nextStateProofRnd, latestRoundInIntervalHdr.Round, votersRound, errExpectedDifferentStateProofRound)
+	}
+
+	acceptableWeight := AcceptableStateProofWeight(*votersHdr, atRound, logging.Base())
```

Comment:
could flip this and have AcceptableStateProofWeight and GetProvenWeight take *BlockHeader